lib/dry/schema/path.rb

```diff
@@ -104,8 +104,19 @@ def <=>(other)
       end
 
       # @api private
-      def key_matches(other)
-        map { |key| (idx = other.index(key)) && keys[idx].equal?(key) }
+      def &(other)
+        unless same_root?(other)
+          raise ArgumentError, "#{other.inspect} doesn't have the same root #{inspect}"
+        end
+
+        self.class.new(
+          key_matches(other, :select).compact.reject { |value| value.equal?(false) }
```

Comment:
is `.compact` needed here? And if so, you could check both conditions in reject, right?